File: 2020/06/Haskell/src/Main.hs
import           Data.List
import           System.Environment
import           Text.Printf

parseInput :: [String] -> [[String]]
parseInput [] = []
parseInput x  = groups : parseInput remaining
 where
  groups    = takeWhile (not . null) x
  remaining = drop 1 . dropWhile (not . null) $ x

intersections :: [String] -> [String]
intersections [x         ] = [x]
intersections (x : y : xs) = intersections $ x `intersect` y : xs
intersections []           = []

part1 :: [String] -> Int
part1 =
  sum . map (length . group . sort . filter (/= ' ') . unwords) . parseInput

part2 :: [String] -> Int
part2 =
  sum
    . map length
    . filter (not . null)
    . map (unwords . intersections)
    . parseInput

output :: String -> IO ()
output path = do
  input <- lines <$> readFile path
  printf "File: %s\n" path
  printf "  Part 1: %d\n" . part1 $ input
  printf "  Part 2: %d\n" . part2 $ input

main :: IO ()
main = getArgs >>= mapM_ output

File: 2020/06/Rust/src/main.rs
use std::collections::HashSet;
use std::env;
use std::fs;

fn parse_input(input: &str) -> Vec<Vec<&str>> {
    let mut vector = Vec::new();
    let mut result = Vec::new();
    for line in input.lines() {
        if line.is_empty() {
            vector.push(result);
            result = vec![];
        } else {
            result.push(line);
        }
    }
    if !result.is_empty() {
        vector.push(result)
    };
    vector
}

fn part_1(input: &[Vec<&str>]) -> usize {
    let mut input = input
        .iter()
        .map(|x| x.concat().chars().collect::<Vec<_>>())
        .collect::<Vec<_>>();
    input.iter_mut().for_each(|x| x.sort_unstable());
    input.iter_mut().for_each(|x| x.dedup());
    input.iter().map(|x| x.len()).sum()
}

fn intersect(input: &[&str]) -> usize {
    let first = input[0];
    let mut set = first.chars().collect::<HashSet<_>>();
    for item in input.iter().skip(1) {
        let next_set = item.chars().collect::<HashSet<_>>();
        set = set.intersection(&next_set).cloned().collect::<HashSet<_>>();
    }
    set.len()
}

fn part_2(input: &[Vec<&str>]) -> usize {
    input.iter().map(|x| intersect(x)).sum()
}

fn output(path: &str) {
    let input = fs::read_to_string(&path).unwrap();
    let parsed = parse_input(&input);

    println!("File: {}", path);
    println!("  Part 1: {}", part_1(&parsed));
    println!("  Part 2: {}", part_2(&parsed));
}

fn main() {
    env::args()
        .skip(1)
        .collect::<Vec<_>>()
        .iter()
        .for_each(|x| output(x));
}
